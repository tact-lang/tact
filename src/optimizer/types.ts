import { AstExpression } from "../grammar/ast";

export abstract class ExpressionTransformer {
    public abstract applyRules(ast: AstExpression): AstExpression;
}

export abstract class Rule {
    public abstract applyRule(
        ast: AstExpression,
        optimizer: ExpressionTransformer,
    ): AstExpression;
}
