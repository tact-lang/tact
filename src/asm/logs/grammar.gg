vmLine =
      VmLoc
    / VmStack
    / VmExecute
    / VmLimitChanged
    / VmGasRemaining
    / VmException
    / VmExceptionHandler
    / VmFinalC5
    / VmUnknown;

VmStack            = "stack: " stack:$[^\n]+;
VmLoc              = "code cell hash:" hash:hex "offset:" offset:number;
VmExecute          = "execute " instr:$[^\n]+;
VmLimitChanged     = "changing gas limit to " limit:number;
VmGasRemaining     = "gas remaining: " gas:number;
VmException        = "handling exception code " errno:number ": " message:$[^\n]*;
VmExceptionHandler = "default exception handler, terminating vm with exit code" errno:number;
VmFinalC5          = "final c5:" value:Cell;
VmUnknown          = !"stack" text:$[^\n]+ "\n";

VmParsedStack = "[" values:VmStackValue* "]";

VmStackValue = value:(
      Null
    / NaN
    / Integer
    / Tuple
    / TupleParen
    / Cell
    / Continuation
    / Builder
    / CellSlice
    / Unknown
);

Null         = "()" / "(null)";
NaN          = "NaN";
Integer      = value:number;
Tuple        = "[" elements:VmStackValue* "]";
TupleParen   = "(" elements:VmStackValue* ")";
Cell         = "C{" value:hex "}";
Continuation = "Cont{" value:$#([A-Za-z_0-9]*) "}";
Builder      = "BC{" value:hex "}";
Unknown      = "???";

CellSlice     = "CS{" body:(CellSliceBody / CellSliceShortBody) "}";
CellSliceBody = "Cell{" value:hex "}" bits:CellSliceBits ";" refs:CellSliceRefs;
CellSliceBits = "bits:" start:number ".." end:number;
CellSliceRefs = "refs:" start:number ".." end:number;

CellSliceShortBody = value:hex;

number   = op:"-"? value:#$(digit+);
digit    = [0-9];

hexDigit "hexadecimal digit" = [0-9a-fA-F];
hex      = $#(value:hexDigit*);

space "space" = [ \t\r\n];
