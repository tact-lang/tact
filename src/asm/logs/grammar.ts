/* Generated. Do not edit. */
/* eslint-disable @typescript-eslint/no-namespace */
/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/no-redundant-type-constituents */
/* eslint-disable @typescript-eslint/no-duplicate-type-constituents */
/* eslint-disable @typescript-eslint/no-unused-vars */
import * as $ from "@tonstudio/parser-runtime"
export namespace $ast {
    export type VmLoc = $.Located<{
        readonly $: "VmLoc"
        readonly hash: hex
        readonly offset: $number
    }>
    export type VmStack = $.Located<{
        readonly $: "VmStack"
        readonly stack: string
    }>
    export type VmExecute = $.Located<{
        readonly $: "VmExecute"
        readonly instr: string
    }>
    export type VmLimitChanged = $.Located<{
        readonly $: "VmLimitChanged"
        readonly limit: $number
    }>
    export type VmGasRemaining = $.Located<{
        readonly $: "VmGasRemaining"
        readonly gas: $number
    }>
    export type VmException = $.Located<{
        readonly $: "VmException"
        readonly errno: $number
        readonly message: string
    }>
    export type VmExceptionHandler = $.Located<{
        readonly $: "VmExceptionHandler"
        readonly errno: $number
    }>
    export type VmFinalC5 = $.Located<{
        readonly $: "VmFinalC5"
        readonly value: Cell
    }>
    export type VmUnknown = $.Located<{
        readonly $: "VmUnknown"
        readonly text: string
    }>
    export type vmLine =
        | VmLoc
        | VmStack
        | VmExecute
        | VmLimitChanged
        | VmGasRemaining
        | VmException
        | VmExceptionHandler
        | VmFinalC5
        | VmUnknown
    export type VmParsedStack = $.Located<{
        readonly $: "VmParsedStack"
        readonly values: readonly VmStackValue[]
    }>
    export type VmStackValue = $.Located<{
        readonly $: "VmStackValue"
        readonly value:
            | Null
            | NaN
            | Integer
            | Tuple
            | TupleParen
            | Cell
            | Continuation
            | Builder
            | CellSlice
            | Unknown
    }>
    export type Null = $.Located<{
        readonly $: "Null"
    }>
    export type NaN = $.Located<{
        readonly $: "NaN"
    }>
    export type Integer = $.Located<{
        readonly $: "Integer"
        readonly value: $number
    }>
    export type Tuple = $.Located<{
        readonly $: "Tuple"
        readonly elements: readonly VmStackValue[]
    }>
    export type TupleParen = $.Located<{
        readonly $: "TupleParen"
        readonly elements: readonly VmStackValue[]
    }>
    export type Cell = $.Located<{
        readonly $: "Cell"
        readonly value: hex
    }>
    export type Continuation = $.Located<{
        readonly $: "Continuation"
        readonly value: string
    }>
    export type Builder = $.Located<{
        readonly $: "Builder"
        readonly value: hex
    }>
    export type Unknown = $.Located<{
        readonly $: "Unknown"
    }>
    export type CellSlice = $.Located<{
        readonly $: "CellSlice"
        readonly body: CellSliceBody | CellSliceShortBody
    }>
    export type CellSliceBody = $.Located<{
        readonly $: "CellSliceBody"
        readonly value: hex
        readonly bits: CellSliceBits
        readonly refs: CellSliceRefs
    }>
    export type CellSliceBits = $.Located<{
        readonly $: "CellSliceBits"
        readonly start: $number
        readonly end: $number
    }>
    export type CellSliceRefs = $.Located<{
        readonly $: "CellSliceRefs"
        readonly start: $number
        readonly end: $number
    }>
    export type CellSliceShortBody = $.Located<{
        readonly $: "CellSliceShortBody"
        readonly value: hex
    }>
    export type $number = {
        readonly op: "-" | undefined
        readonly value: string
    }
    export type digit = string
    export type hexDigit = string | string | string
    export type hex = string
    export type space = " " | "\t" | "\r" | "\n"
}
export const VmLoc: $.Parser<$ast.VmLoc> = $.loc(
    $.field(
        $.pure("VmLoc"),
        "$",
        $.right(
            $.str("code cell hash:"),
            $.field(
                $.lazy(() => hex),
                "hash",
                $.right(
                    $.str("offset:"),
                    $.field(
                        $.lazy(() => $number),
                        "offset",
                        $.eps,
                    ),
                ),
            ),
        ),
    ),
)
export const VmStack: $.Parser<$ast.VmStack> = $.loc(
    $.field(
        $.pure("VmStack"),
        "$",
        $.right(
            $.str("stack: "),
            $.field(
                $.stry($.plus($.regex<"\n">("^\\n", $.negateExps([$.ExpString("\n")])))),
                "stack",
                $.eps,
            ),
        ),
    ),
)
export const VmExecute: $.Parser<$ast.VmExecute> = $.loc(
    $.field(
        $.pure("VmExecute"),
        "$",
        $.right(
            $.str("execute "),
            $.field(
                $.stry($.plus($.regex<"\n">("^\\n", $.negateExps([$.ExpString("\n")])))),
                "instr",
                $.eps,
            ),
        ),
    ),
)
export const VmLimitChanged: $.Parser<$ast.VmLimitChanged> = $.loc(
    $.field(
        $.pure("VmLimitChanged"),
        "$",
        $.right(
            $.str("changing gas limit to "),
            $.field(
                $.lazy(() => $number),
                "limit",
                $.eps,
            ),
        ),
    ),
)
export const VmGasRemaining: $.Parser<$ast.VmGasRemaining> = $.loc(
    $.field(
        $.pure("VmGasRemaining"),
        "$",
        $.right(
            $.str("gas remaining: "),
            $.field(
                $.lazy(() => $number),
                "gas",
                $.eps,
            ),
        ),
    ),
)
export const VmException: $.Parser<$ast.VmException> = $.loc(
    $.field(
        $.pure("VmException"),
        "$",
        $.right(
            $.str("handling exception code "),
            $.field(
                $.lazy(() => $number),
                "errno",
                $.right(
                    $.str(": "),
                    $.field(
                        $.stry($.star($.regex<"\n">("^\\n", $.negateExps([$.ExpString("\n")])))),
                        "message",
                        $.eps,
                    ),
                ),
            ),
        ),
    ),
)
export const VmExceptionHandler: $.Parser<$ast.VmExceptionHandler> = $.loc(
    $.field(
        $.pure("VmExceptionHandler"),
        "$",
        $.right(
            $.str("default exception handler, terminating vm with exit code"),
            $.field(
                $.lazy(() => $number),
                "errno",
                $.eps,
            ),
        ),
    ),
)
export const VmFinalC5: $.Parser<$ast.VmFinalC5> = $.loc(
    $.field(
        $.pure("VmFinalC5"),
        "$",
        $.right(
            $.str("final c5:"),
            $.field(
                $.lazy(() => Cell),
                "value",
                $.eps,
            ),
        ),
    ),
)
export const VmUnknown: $.Parser<$ast.VmUnknown> = $.loc(
    $.field(
        $.pure("VmUnknown"),
        "$",
        $.right(
            $.lookNeg($.str("stack")),
            $.field(
                $.stry($.plus($.regex<"\n">("^\\n", $.negateExps([$.ExpString("\n")])))),
                "text",
                $.right($.str("\n"), $.eps),
            ),
        ),
    ),
)
export const vmLine: $.Parser<$ast.vmLine> = $.alt(
    VmLoc,
    $.alt(
        VmStack,
        $.alt(
            VmExecute,
            $.alt(
                VmLimitChanged,
                $.alt(
                    VmGasRemaining,
                    $.alt(VmException, $.alt(VmExceptionHandler, $.alt(VmFinalC5, VmUnknown))),
                ),
            ),
        ),
    ),
)
export const VmParsedStack: $.Parser<$ast.VmParsedStack> = $.loc(
    $.field(
        $.pure("VmParsedStack"),
        "$",
        $.right(
            $.str("["),
            $.field($.star($.lazy(() => VmStackValue)), "values", $.right($.str("]"), $.eps)),
        ),
    ),
)
export const VmStackValue: $.Parser<$ast.VmStackValue> = $.loc(
    $.field(
        $.pure("VmStackValue"),
        "$",
        $.field(
            $.alt(
                $.lazy(() => Null),
                $.alt(
                    $.lazy(() => NaN),
                    $.alt(
                        $.lazy(() => Integer),
                        $.alt(
                            $.lazy(() => Tuple),
                            $.alt(
                                $.lazy(() => TupleParen),
                                $.alt(
                                    $.lazy(() => Cell),
                                    $.alt(
                                        $.lazy(() => Continuation),
                                        $.alt(
                                            $.lazy(() => Builder),
                                            $.alt(
                                                $.lazy(() => CellSlice),
                                                $.lazy(() => Unknown),
                                            ),
                                        ),
                                    ),
                                ),
                            ),
                        ),
                    ),
                ),
            ),
            "value",
            $.eps,
        ),
    ),
)
export const Null: $.Parser<$ast.Null> = $.loc(
    $.field($.pure("Null"), "$", $.right($.alt($.str("()"), $.str("(null)")), $.eps)),
)
export const NaN: $.Parser<$ast.NaN> = $.loc(
    $.field($.pure("NaN"), "$", $.right($.str("NaN"), $.eps)),
)
export const Integer: $.Parser<$ast.Integer> = $.loc(
    $.field(
        $.pure("Integer"),
        "$",
        $.field(
            $.lazy(() => $number),
            "value",
            $.eps,
        ),
    ),
)
export const Tuple: $.Parser<$ast.Tuple> = $.loc(
    $.field(
        $.pure("Tuple"),
        "$",
        $.right($.str("["), $.field($.star(VmStackValue), "elements", $.right($.str("]"), $.eps))),
    ),
)
export const TupleParen: $.Parser<$ast.TupleParen> = $.loc(
    $.field(
        $.pure("TupleParen"),
        "$",
        $.right($.str("("), $.field($.star(VmStackValue), "elements", $.right($.str(")"), $.eps))),
    ),
)
export const Cell: $.Parser<$ast.Cell> = $.loc(
    $.field(
        $.pure("Cell"),
        "$",
        $.right(
            $.str("C{"),
            $.field(
                $.lazy(() => hex),
                "value",
                $.right($.str("}"), $.eps),
            ),
        ),
    ),
)
export const Continuation: $.Parser<$ast.Continuation> = $.loc(
    $.field(
        $.pure("Continuation"),
        "$",
        $.right(
            $.str("Cont{"),
            $.field(
                $.stry(
                    $.lex(
                        $.star(
                            $.regex<string | string | "_" | string>("A-Za-z_0-9", [
                                $.ExpRange("A", "Z"),
                                $.ExpRange("a", "z"),
                                $.ExpString("_"),
                                $.ExpRange("0", "9"),
                            ]),
                        ),
                    ),
                ),
                "value",
                $.right($.str("}"), $.eps),
            ),
        ),
    ),
)
export const Builder: $.Parser<$ast.Builder> = $.loc(
    $.field(
        $.pure("Builder"),
        "$",
        $.right(
            $.str("BC{"),
            $.field(
                $.lazy(() => hex),
                "value",
                $.right($.str("}"), $.eps),
            ),
        ),
    ),
)
export const Unknown: $.Parser<$ast.Unknown> = $.loc(
    $.field($.pure("Unknown"), "$", $.right($.str("???"), $.eps)),
)
export const CellSlice: $.Parser<$ast.CellSlice> = $.loc(
    $.field(
        $.pure("CellSlice"),
        "$",
        $.right(
            $.str("CS{"),
            $.field(
                $.alt(
                    $.lazy(() => CellSliceBody),
                    $.lazy(() => CellSliceShortBody),
                ),
                "body",
                $.right($.str("}"), $.eps),
            ),
        ),
    ),
)
export const CellSliceBody: $.Parser<$ast.CellSliceBody> = $.loc(
    $.field(
        $.pure("CellSliceBody"),
        "$",
        $.right(
            $.str("Cell{"),
            $.field(
                $.lazy(() => hex),
                "value",
                $.right(
                    $.str("}"),
                    $.field(
                        $.lazy(() => CellSliceBits),
                        "bits",
                        $.right(
                            $.str(";"),
                            $.field(
                                $.lazy(() => CellSliceRefs),
                                "refs",
                                $.eps,
                            ),
                        ),
                    ),
                ),
            ),
        ),
    ),
)
export const CellSliceBits: $.Parser<$ast.CellSliceBits> = $.loc(
    $.field(
        $.pure("CellSliceBits"),
        "$",
        $.right(
            $.str("bits:"),
            $.field(
                $.lazy(() => $number),
                "start",
                $.right(
                    $.str(".."),
                    $.field(
                        $.lazy(() => $number),
                        "end",
                        $.eps,
                    ),
                ),
            ),
        ),
    ),
)
export const CellSliceRefs: $.Parser<$ast.CellSliceRefs> = $.loc(
    $.field(
        $.pure("CellSliceRefs"),
        "$",
        $.right(
            $.str("refs:"),
            $.field(
                $.lazy(() => $number),
                "start",
                $.right(
                    $.str(".."),
                    $.field(
                        $.lazy(() => $number),
                        "end",
                        $.eps,
                    ),
                ),
            ),
        ),
    ),
)
export const CellSliceShortBody: $.Parser<$ast.CellSliceShortBody> = $.loc(
    $.field(
        $.pure("CellSliceShortBody"),
        "$",
        $.field(
            $.lazy(() => hex),
            "value",
            $.eps,
        ),
    ),
)
export const $number: $.Parser<$ast.$number> = $.field(
    $.opt($.str("-")),
    "op",
    $.field($.lex($.stry($.plus($.lazy(() => digit)))), "value", $.eps),
)
export const digit: $.Parser<$ast.digit> = $.regex<string>("0-9", [$.ExpRange("0", "9")])
export const hexDigit: $.Parser<$ast.hexDigit> = $.named(
    "hexadecimal digit",
    $.regex<string | string | string>("0-9a-fA-F", [
        $.ExpRange("0", "9"),
        $.ExpRange("a", "f"),
        $.ExpRange("A", "F"),
    ]),
)
export const hex: $.Parser<$ast.hex> = $.stry($.lex($.field($.star(hexDigit), "value", $.eps)))
export const space: $.Parser<$ast.space> = $.named(
    "space",
    $.regex<" " | "\t" | "\r" | "\n">(" \\t\\r\\n", [
        $.ExpString(" "),
        $.ExpString("\t"),
        $.ExpString("\r"),
        $.ExpString("\n"),
    ]),
)
