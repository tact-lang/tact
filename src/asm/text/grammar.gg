File = instructions:instructions;

instructions = (Instruction / ExplicitRef / EmbedSlice / Exotic)*;

Instruction = name:Id args:Argument*;

Argument = expression:(
      IntegerLiteral
    / DataLiteral
    / Code
    / Dictionary
    / StackElement
    / ControlRegister
);

StackElement = value:$#("s" "-"? IntegerLiteralDec);
ControlRegister = value:$#("c" IntegerLiteralDec);

Code = "{" instructions:instructions "}";

Dictionary = "[" entries:DictionaryEntry* "]";

DictionaryEntry = id:IntegerLiteral "=>" code:Code;

ExplicitRef = keyword<"ref"> code:Code;
EmbedSlice  = keyword<"embed"> data:DataLiteral;
Exotic      = keyword<"exotic"> lib:(ExoticLibrary / DefaultExotic);

ExoticLibrary = keyword<"library"> data:DataLiteral;
DefaultExotic = data:DataLiteral;

idPart "identifier character" = [a-zA-Z0-9_];
Id "identifier" = name:#$(!reservedWord [a-zA-Z_] idPart*);

DataLiteral = value:(HexLiteral / BinLiteral / BocLiteral / StringLiteral);

HexLiteral = "x{" content:$(hexDigit* "_"?) "}";
BinLiteral = "b{" content:$(binDigit*) "}";
BocLiteral = "boc{" content:$(hexDigit*) "}";

// Literals
IntegerLiteral = op:"-"? value:(IntegerLiteralHex / IntegerLiteralBin / IntegerLiteralOct / IntegerLiteralDec);

IntegerLiteralDec = digits:#underscored<digit>;
IntegerLiteralHex = digits:#("0" [xX] @underscored<hexDigit>);
IntegerLiteralBin = digits:#("0" [bB] @underscored<binDigit>);
IntegerLiteralOct = digits:#("0" [oO] @underscored<[0-7]>);

underscored<T> = $(T ("_"? T)*);

hexDigit "hexadecimal digit" = [0-9a-fA-F];
binDigit "binary digit" = [01];
digit "digit" = [0-9];

StringLiteral = value:#("\"" @$([^"\\])* "\"");

keyword<T> = #(@T !idPart);

reservedWord "reserved word" = keyword<(
    "ref" / "embed" / "exotic" / "library"
)>;

space "space" = [ \t\r\n] / comment;
comment = singleLineComment;
singleLineComment = "//" @$[^\r\n]*;
