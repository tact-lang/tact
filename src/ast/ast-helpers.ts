import * as A from "./ast";
import { AstId } from "./ast";
import { throwInternalCompilerError } from "../error/errors";
import { dummySrcInfo } from "../grammar";

/**
 * Check if input expression is a 'path expression',
 * i.e. an identifier or a sequence of field accesses starting from an identifier.
 * @param path A path expression to check.
 * @returns An array of identifiers or null if the input expression is not a path expression.
 */
export function tryExtractPath(path: A.AstExpression): A.AstId[] | null {
    switch (path.kind) {
        case "id":
            return [path];
        case "field_access": {
            const p = tryExtractPath(path.aggregate);
            return p ? [...p, path.field] : null;
        }
        default:
            return null;
    }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type DistributiveOmit<T, K extends keyof any> = T extends any
    ? Omit<T, K>
    : never;

export const getAstFactory = () => {
    let nextId = 1;
    function createNode(src: DistributiveOmit<A.AstNode, "id">): A.AstNode {
        return Object.freeze(Object.assign({ id: nextId++ }, src));
    }
    function cloneNode<T extends A.AstNode>(src: T): T {
        const newNode: T = { ...src, id: nextId++ };
        return Object.freeze(newNode);
    }
    return {
        createNode,
        cloneNode,
    };
};

export type FactoryAst = ReturnType<typeof getAstFactory>;

export function idText(ident: A.AstId | A.AstFuncId | A.AstTypeId): string {
    return ident.text;
}

export function isInt(ident: A.AstTypeId): boolean {
    return ident.text === "Int";
}

export function isBool(ident: A.AstTypeId): boolean {
    return ident.text === "Bool";
}

export function isCell(ident: A.AstTypeId): boolean {
    return ident.text === "Cell";
}

export function isSlice(ident: A.AstTypeId): boolean {
    return ident.text === "Slice";
}

export function isBuilder(ident: A.AstTypeId): boolean {
    return ident.text === "Builder";
}

export function isAddress(ident: A.AstTypeId): boolean {
    return ident.text === "Address";
}

export function isString(ident: A.AstTypeId): boolean {
    return ident.text === "String";
}

export function isStringBuilder(ident: A.AstTypeId): boolean {
    return ident.text === "StringBuilder";
}

export function isSelfId(ident: A.AstId): boolean {
    return ident.text === "self";
}

export function isWildcard(ident: A.AstId): boolean {
    return ident.text === "_";
}

export function isRequire(ident: A.AstId): boolean {
    return ident.text === "require";
}

export function eqNames(
    left: A.AstId | A.AstTypeId | string,
    right: A.AstId | A.AstTypeId | string,
): boolean {
    if (typeof left === "string") {
        if (typeof right === "string") {
            return left === right;
        }
        return left === right.text;
    } else {
        if (typeof right === "string") {
            return left.text === right;
        }
        return left.text === right.text;
    }
}

export function idOfText(text: string): A.AstId {
    return {
        kind: "id",
        text,
        id: 0,
        loc: dummySrcInfo,
    };
}

export function astNumToString(n: A.AstNumber): string {
    switch (n.base) {
        case 2:
            return `0b${n.value.toString(n.base)}`;
        case 8:
            return `0o${n.value.toString(n.base)}`;
        case 10:
            return n.value.toString(n.base);
        case 16:
            return `0x${n.value.toString(n.base)}`;
    }
}

// Test equality of AstExpressions.
// Note this is syntactical equality of expressions.
// For example, two struct instances are equal if they have the same
// type and same fields in the same order.
export function eqExpressions(
    ast1: A.AstExpression,
    ast2: A.AstExpression,
): boolean {
    if (ast1.kind !== ast2.kind) {
        return false;
    }

    switch (ast1.kind) {
        case "null":
            return true;
        case "boolean":
            return ast1.value === (ast2 as A.AstBoolean).value;
        case "number":
            return ast1.value === (ast2 as A.AstNumber).value;
        case "string":
            return ast1.value === (ast2 as A.AstString).value;
        case "id":
            return eqNames(ast1, ast2 as A.AstId);
        case "address":
            return ast1.value.equals((ast2 as A.AstAddress).value);
        case "cell":
            return ast1.value.equals((ast2 as A.AstCell).value);
        case "slice":
            return ast1.value
                .asCell()
                .equals((ast2 as A.AstSlice).value.asCell());
        case "simplified_string":
            return ast1.value === (ast2 as A.AstSimplifiedString).value;
        case "struct_value":
            return (
                eqNames(ast1.type, (ast2 as A.AstStructValue).type) &&
                eqArrays(
                    ast1.args,
                    (ast2 as A.AstStructValue).args,
                    eqFieldValues,
                )
            );
        case "method_call":
            return (
                eqNames(ast1.method, (ast2 as A.AstMethodCall).method) &&
                eqExpressions(ast1.self, (ast2 as A.AstMethodCall).self) &&
                eqArrays(
                    ast1.args,
                    (ast2 as A.AstMethodCall).args,
                    eqExpressions,
                )
            );
        case "init_of":
            return (
                eqNames(ast1.contract, (ast2 as A.AstInitOf).contract) &&
                eqArrays(ast1.args, (ast2 as A.AstInitOf).args, eqExpressions)
            );
        case "op_unary":
            return (
                ast1.op === (ast2 as A.AstOpUnary).op &&
                eqExpressions(ast1.operand, (ast2 as A.AstOpUnary).operand)
            );
        case "op_binary":
            return (
                ast1.op === (ast2 as A.AstOpBinary).op &&
                eqExpressions(ast1.left, (ast2 as A.AstOpBinary).left) &&
                eqExpressions(ast1.right, (ast2 as A.AstOpBinary).right)
            );
        case "conditional":
            return (
                eqExpressions(
                    ast1.condition,
                    (ast2 as A.AstConditional).condition,
                ) &&
                eqExpressions(
                    ast1.thenBranch,
                    (ast2 as A.AstConditional).thenBranch,
                ) &&
                eqExpressions(
                    ast1.elseBranch,
                    (ast2 as A.AstConditional).elseBranch,
                )
            );
        case "struct_instance":
            return (
                eqNames(ast1.type, (ast2 as A.AstStructInstance).type) &&
                eqArrays(
                    ast1.args,
                    (ast2 as A.AstStructInstance).args,
                    eqFieldInitializers,
                )
            );
        case "field_access":
            return (
                eqNames(ast1.field, (ast2 as A.AstFieldAccess).field) &&
                eqExpressions(
                    ast1.aggregate,
                    (ast2 as A.AstFieldAccess).aggregate,
                )
            );
        case "static_call":
            return (
                eqNames(ast1.function, (ast2 as A.AstStaticCall).function) &&
                eqArrays(
                    ast1.args,
                    (ast2 as A.AstStaticCall).args,
                    eqExpressions,
                )
            );
        default:
            throwInternalCompilerError("Unrecognized expression kind");
    }
}

function eqFieldInitializers(
    arg1: A.AstStructFieldInitializer,
    arg2: A.AstStructFieldInitializer,
): boolean {
    return (
        eqNames(arg1.field, arg2.field) &&
        eqExpressions(arg1.initializer, arg2.initializer)
    );
}

function eqFieldValues(
    arg1: A.AstStructFieldValue,
    arg2: A.AstStructFieldValue,
): boolean {
    return (
        eqNames(arg1.field, arg2.field) &&
        eqExpressions(arg1.initializer, arg2.initializer)
    );
}

function eqArrays<T>(
    arr1: readonly T[],
    arr2: readonly T[],
    eqElements: (elem1: T, elem2: T) => boolean,
): boolean {
    if (arr1.length !== arr2.length) {
        return false;
    }

    for (let i = 0; i < arr1.length; i++) {
        if (!eqElements(arr1[i]!, arr2[i]!)) {
            return false;
        }
    }

    return true;
}

/*
Functions that return guard types like "ast is AstLiteral" are unsafe to use in production code.
But there is a way to make them safe by introducing an intermediate function, like
the "checkLiteral" function defined below after "isLiteral". In principle, it is possible to use "checkLiteral"
directly in the code (which avoids the guard type altogether), but it produces code that reduces readability significantly.

The pattern shown with "isLiteral" and "checkLiteral" can be generalized to other functions that produce a guard type
based on a decision of several cases.
For example, if we have the following function, where we assume that B is a subtype of A:

function isB(d: A): d is B {
  if (cond1(d)) {        // It is assumed that cond1(d) determines d to be of type B inside the if
     return true;
  } else if (cond2(d)) { // It is assumed that cond2(d) determines d to be of type A but not of type B inside the if
     return false;
  } else if (cond3(d)) { // It is assumed that cond3(d) determines d to be of type B inside the if
     return true;
  } else {               // It is assumed that d is of type A but not of type B inside the else
     return false;
  }
}

We can introduce a "checkB" function as follows:

function checkB<T>(d: A, t: (arg: B) => T, f: (arg: Exclude<A,B>) => T): T {
  if (cond1(d)) {
     return t(d);
  } else if (cond2(d)) {
     return f(d);
  } else if (cond3(d)) {
     return t(d);
  } else {
     return f(d);
  }
}

Here, all the "true" cases return t(d) and all the "false" cases return f(d). The names of the functions t and f help remember
that they correspond to the true and false cases, respectively. Observe that cond1(d) and cond3(d) determine the type of
d to be B, which means we can pass d to the t function. For the false cases, the type of d is determined to be
A but not B, which means we can pass d to function f, because f's argument type Exclude<A,B> states
that the argument must be of type A but not of type B, i.e., of type "A - B" if we see the types as sets.

checkB is safe because the compiler will complain if, for example, we use t(d) in the else case:

function checkB<T>(d: A, t: (arg: B) => T, f: (arg: Exclude<A,B>) => T): T {
  if (cond1(d)) {
     return t(d);
  } else if (cond2(d)) {
     return f(d);
  } else if (cond3(d)) {
     return t(d);
  } else {
     return t(d);   // Compiler will signal an error that d is not assignable to type B
  }
}

Contrary to the original function, where the compiler remains silent if we incorrectly return true in the else:

function isB(d: A): d is B {
  if (cond1(d)) {
     return true;
  } else if (cond2(d)) {
     return false;
  } else if (cond3(d)) {
     return true;
  } else {
     return true;   // Wrong, but compiler remains silent
  }
}

After we have our "checkB" function, we can define the "isB" function simply as:

function isB(d: A): d is B {
  return checkB(d, () => true, () => false);
}
*/

export function isLiteral(ast: A.AstExpression): ast is A.AstLiteral {
    return checkLiteral(
        ast,
        () => true,
        () => false,
    );
}

function checkLiteral<T>(
    ast: A.AstExpression,
    t: (node: A.AstLiteral) => T,
    f: (node: Exclude<A.AstExpression, A.AstLiteral>) => T,
): T {
    switch (ast.kind) {
        case "null":
        case "boolean":
        case "number":
        case "address":
        case "cell":
        case "slice":
        case "simplified_string":
        case "struct_value":
            return t(ast);

        case "struct_instance":
        case "string":
        case "id":
        case "method_call":
        case "init_of":
        case "op_unary":
        case "op_binary":
        case "conditional":
        case "field_access":
        case "static_call":
            return f(ast);

        default:
            throwInternalCompilerError("Unrecognized expression kind");
    }
}

export const selfId: AstId = {
    kind: "id",
    text: "self",
    id: 0,
    loc: dummySrcInfo,
};
