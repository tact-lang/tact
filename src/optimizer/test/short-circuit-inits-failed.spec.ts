import { getAstFactory } from "../../ast/ast-helpers";
import { CompilerContext } from "../../context/context";
import { openContext } from "../../context/store";
import { getParser } from "../../grammar";
import { defaultParser } from "../../grammar/grammar";
import { resolveDescriptors } from "../../types/resolveDescriptors";
import { resolveSignatures } from "../../types/resolveSignatures";
import { resolveStatements } from "../../types/resolveStatements";
import { loadCases } from "../../utils/loadCases";

describe("short-circuit-initialization", () => {
    for (const r of loadCases(__dirname + "/failed/")) {
        it(`${r.name} should fail boolean short-circuiting during initialization`, () => {
            const Ast = getAstFactory();
            let ctx = openContext(
                new CompilerContext(),
                [{ code: r.code, path: "<unknown>", origin: "user" }],
                [],
                getParser(Ast, defaultParser),
            );
            expect(() => {
                ctx = resolveDescriptors(ctx, Ast);
                ctx = resolveStatements(ctx, Ast);
                ctx = resolveSignatures(ctx, Ast);
            }).toThrowErrorMatchingSnapshot();
        });
    }
});
