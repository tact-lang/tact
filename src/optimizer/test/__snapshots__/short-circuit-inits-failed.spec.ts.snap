// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`short-circuit-initialization short-circuit-constant-init-and should fail boolean short-circuiting during initialization 1`] = `
"<unknown>:4:35: Cannot evaluate expression to a constant: divisor expression must be non-zero
  3 | 
> 4 | const A: Bool = true && exception(0);
                                        ^
  5 | 
"
`;

exports[`short-circuit-initialization short-circuit-constant-init-or should fail boolean short-circuiting during initialization 1`] = `
"<unknown>:4:36: Cannot evaluate expression to a constant: divisor expression must be non-zero
  3 | 
> 4 | const A: Bool = false || exception(0);
                                         ^
  5 | 
"
`;

exports[`short-circuit-initialization short-circuit-contract-constant-init-and should fail boolean short-circuiting during initialization 1`] = `
"<unknown>:12:39: Cannot evaluate expression to a constant: divisor expression must be non-zero
  11 | contract CTest {
> 12 |     const A: Bool = true && exception(0);
                                             ^
  13 | }
"
`;

exports[`short-circuit-initialization short-circuit-contract-constant-init-or should fail boolean short-circuiting during initialization 1`] = `
"<unknown>:12:40: Cannot evaluate expression to a constant: divisor expression must be non-zero
  11 | contract CTest {
> 12 |     const A: Bool = false || exception(0);
                                              ^
  13 | }
"
`;

exports[`short-circuit-initialization short-circuit-contract-init-and should fail boolean short-circuiting during initialization 1`] = `
"<unknown>:12:33: Cannot evaluate expression to a constant: divisor expression must be non-zero
  11 | contract CTest {
> 12 |     A: Bool = true && exception(0);
                                       ^
  13 | }
"
`;

exports[`short-circuit-initialization short-circuit-contract-init-or should fail boolean short-circuiting during initialization 1`] = `
"<unknown>:12:34: Cannot evaluate expression to a constant: divisor expression must be non-zero
  11 | contract CTest {
> 12 |     A: Bool = false || exception(0);
                                        ^
  13 | }
"
`;

exports[`short-circuit-initialization short-circuit-struct-init-and should fail boolean short-circuiting during initialization 1`] = `
"<unknown>:5:33: Cannot evaluate expression to a constant: divisor expression must be non-zero
  4 | struct STest {
> 5 |     A: Bool = true && exception(0);
                                      ^
  6 | }
"
`;

exports[`short-circuit-initialization short-circuit-struct-init-or should fail boolean short-circuiting during initialization 1`] = `
"<unknown>:5:34: Cannot evaluate expression to a constant: divisor expression must be non-zero
  4 | struct STest {
> 5 |     A: Bool = false || exception(0);
                                       ^
  6 | }
"
`;
